src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -623,6 +628,28 @@ private BoundLambda ReallyInferReturnType(NamedTypeSymbol delegateType, Immutabl
             }
             lambdaSymbol.SetInferredReturnType(inferredReturnType.RefKind, returnType);
 
+            if (forErrorRecovery)
+            {
+                // When binding the lambda for error recovery, we bind one more time to permit the return expressions
+                // to be converted to the inferred return type so that no unconverted expressions remain in the bound tree.
+                diagnostics = DiagnosticBag.GetInstance();
+                lambdaSymbol = new LambdaSymbol(
+                    Binder.Compilation,
+                    Binder.ContainingMemberOrLambda,
+                    _unboundLambda,
+                    parameterTypes,
+                    parameterRefKinds,
+                    refKind: CodeAnalysis.RefKind.None,
+                    returnType: returnType,
+                    diagnostics: diagnostics);
+                lambdaBodyBinder = new ExecutableCodeBinder(_unboundLambda.Syntax, lambdaSymbol, ParameterBinder(lambdaSymbol, Binder));
+                block = BindLambdaBody(lambdaSymbol, lambdaBodyBinder, diagnostics);
```

Comment:
>block = BindLambdaBody(lambdaSymbol, lambdaBodyBinder, diagnostics); [](start = 16, length = 68)

It looks like this statement and the two statements above is almost a copy of  similar lines above. Consider extracting them into a helper and reusing in both places.